README.markdown

```diff
@@ -1054,7 +1069,42 @@ messageView.color = .messageColor
 messageView.color = UIColor(red: 0.121568627, green: 0.129411765, blue: 0.141176471, alpha: 1)
 ```
 
+## Unit Tests
+
+When writing new code, it should have a considerable amount of tests backing those additions. There is no code coverage requisite yet for this.
+
+It's been agreed that unit tests should be written using the Quick/Nimble framework because of the increased readability over `XCTest`. The methods `describe`, `context` and `it` should be used accordingly to build the test specs in a manner that is easy to read, like english, such as:
+```swift
+describe("Given a set of strings") {
+    context("When comparing two strings that are equal") {
+        it("Should return true") {
+            /* Test code goes here */
+        }
+    }
+}
+```
+
+When writing your tests use force unwraps (`!`) to unwrap optional values when needed since safely unwraping (`?` or using `guard`/`if` statements) may give us false positives (not fail a test when it should). The only acceptable case to not forcibly unwrap a variable is when it is being compared to another value using the `expect` method, since if it is `nil` the test will fail with a nice test error message instead of a stacktrace error. E.g.
+```swift
+struct SomeStruct {
+    let name: String
+}
+
+let optionalValue: SomeStruct? = SomeStruct(name: "test")
+expect(optionalValue?.name).to(equal("test")) // This will fail with a message saying value was nil if optionalValue is nil rather than blowing everything up like using !
+```
+Please notice that one of the sides of the comparison must not be optional, otherwise we may end up with a false positive, e.g.
+```swift
+let dateFormatter = DateFormatter()
+dateFormatter.dateFormat = "yyyy-MM-dd"
+
+struct SomeStruct {
+    let date: Date
+}
 
+let optionalValue: SomeStruct? = SomeStruct(date: Date())
+expect(optionalValue?.date).to(equal(dateFormatter.date(from: "2018-06-20")!)) // Notice than one of the sides must be forcibly unwraped otherwise we might be comparing two nil objects, which might not be what we want. Force unwrap your test values (the desired value), not the values to be tested
+```
 
```

Comment:
I suggest we state that one test case should test just one thing. I have seen some tests where on the same test we test more than one (see `DataProviderSpec`).

Reply:
Honestly, I don't agree, I feel if we do only one comparison per test case we are going to have a lot of overhead, I believe the comparisons should be consistent with the test case and try not to test a lot of things in one single test case, but saying that a test case should have one comparison is too extreme, we'll easily end up with test file over 1000 lines

Reply:
DataProviderSpec may be a bit special. But in general terms, when you do TDD you tent to go one thing at a time and then, the end result is one test for one thing without overhead.

That said a Guideline, is just that a guideline that we try to follow and that we may violate in some cases writing the proper comment without major problems.

Reply:
You might not want to look at the `DataTests` as well 😆

Reply:
Ha! As I said it is just a recommendation.